src/wui/attack_box.cc

```diff
@@ -91,24 +94,35 @@ UI::Textarea& AttackBox::add_text(UI::Box& parent,
 	return result;
 }
 
-std::unique_ptr<UI::Button> AttackBox::add_button(UI::Box& parent,
-                                                  const std::string& text,
-                                                  void (AttackBox::*fn)(),
-                                                  const std::string& tooltip_text) {
-	std::unique_ptr<UI::Button> button(new UI::Button(
-	   &parent, text, 8, 8, 34, 34, UI::ButtonStyle::kWuiPrimary, text, tooltip_text));
-	button->sigclicked.connect([this, fn]() { (this->*fn)(); });
-	parent.add(button.get());
+template <typename T>
+UI::Button* add_button(AttackBox* a,
+                       UI::Box& parent,
+                       const std::string& name,
+                       const T& text_or_image,
+                       void (AttackBox::*fn)(),
+                       UI::ButtonStyle style,
+                       const std::string& tooltip_text) {
+	UI::Button* button =
+	   new UI::Button(&parent, name, 8, 8, 34, 34, style, text_or_image, tooltip_text);
+	button->sigclicked.connect([a, fn]() { (a->*fn)(); });
+	parent.add(button);
 	return button;
 }
 
 /*
  * Update available soldiers
  */
 void AttackBox::think() {
-	if ((player_->egbase().get_gametime() - lastupdate_) > kUpdateTimeInGametimeMs) {
+	if (!iplayer_.player().is_seeing(iplayer_.egbase().map().get_index(node_coordinates_)) &&
+	    !iplayer_.player().see_all()) {
+		die();
+	}
+
+	if ((iplayer_.egbase().get_gametime() - lastupdate_) > kUpdateTimeInGametimeMs) {
 		update_attack(false);
```

Comment:
I think update_attack could be renamed - maybe update_window or update_content
Also the class should be renamed since it is a window now and not a box anymore ;)